### ISSUE TITLE
Hilfe

### ISSUE THREAD

**Issue by Helix-FX** [hdo/pico-estlcam-handrad#1]
Hi, 

ich bin leider absoluter Anfänger mit dem Thema Pi / Pico Pi. 
Kannst du mir kurz erklären wie ich die Programmierung auf den Pico Pi bekomme? Ich habe Thonny schon auf meinem Windows PC installiert und auch schon "Hello World" und Blink an den Pico Pi gesendet. Aber alles was ich hier sehe übersteigt leider meine Kompetenz :( 

Könntest du zusätzlich noch eine Zeichnung bei Fritzing o.Ä. machen? Damit ich genau sehen kann welches Kabel vom 4 Achsen Joystick wo hin gehen muss usw.?

Ich hoffe du hast Zeit und Lust mir bei dem Projekt zu helfen....das wäre genial.

Schöne Grüße,
Helix

**Comment by hdo** [comment]
Hi Helix,

für die Programmierung brauchst du nicht extra Thonny installieren, da die Firmware bereits kompiliert ist und direkt aufgespielt werden kann.

Das Aufspielen geht über Drag und Drop. Dafür musst du den PICO in den Programmier-Modus bringen. Den Knopf auf dem PICO gedrückt halten und dann erst das USB Kabel verbinden. Unter Windows wirst du dann ein neues Laufwerk sehen. Die Firmware (UF2 Datei) kannst du dann einfach auf das Laufwerk kopieren, danach ist die Firmware aufgespielt.

Der Joystick hat 3 Analoge Potis, X,Y und Z. Die X und Y Potis kannst du ja sehen. Der Z-POTI wird über die Kabel rausgeführt, da ist die Belegung je nach Bezug unterschiedlich, am besten müsstest du mal die Kabel mit einem Durchgangsprüfer durchmessen. Wenn du einen Knopf auf dem Joystick hast, müssten es 4 Kabel sein:

GND, VCC, ADC und BTN

Die Potis werden wie folgt angeschlossen: GND an GND, VCC an die 3.3V Leitung, ADC an den Eingang des PICO.

ADC ist der mittlere Anschluss.


![2022-01-06_09-51-28](https://user-images.githubusercontent.com/2122230/148356724-bd6be497-ff15-4c7c-9433-e2655fc7e4ff.png)

Ich kenne mich mit Fritzing nicht aus, evtl. müsste ich mir das mal anschauen.

**Comment by Helix-FX** [comment]
Sorry, ich verstehe irgendwie nicht wie ich das was du programmiert hast auf den Pico Pi bekomme oder muss das gar nicht auf den Pi?

Ok, das mit dem Joystick und den Buttons bekomme ich glaube ich hin. 
Die Buttons müssen ja immer auf Masse und Beispielsweise GP13 für den Yellow Button geschaltet werden oder?

Vielen Dank!

**Comment by hdo** [comment]
Doch, die Firmware muss schon auf den PICO ;-)

Wenn du den Knopf gedrückt hältst und dann erst den PICO an den PC anschließt, bekommst du z.B. so eine Ansicht:

![2022-01-06_10-52-37](https://user-images.githubusercontent.com/2122230/148364441-0379db5c-8a20-4789-bc95-176843c02981.png)

Dort einfach die UF2 Datei hin kopieren, fertig ;-)

>> Die Buttons müssen ja immer auf Masse und Beispielsweise GP13 für den Yellow Button geschaltet werden oder?

Genau.

Wobei nur 3 Buttons tatsächlich von Estlcam verwendet werden: OK, FEED, Spindle.

Ich habe das bei mir wie folgt:

OK: Joystick Button
FEED: Grün
Spindle: Ich glaube bei Spindle Drehencoder

Rot und Gelb sind (noch) nicht belegt, da das Protokoll damit nichts anfangen kann.
Ich wollte die beiden Buttons für Macros verwenden, sobald die Funktion bei Estlcam drin ist.


EDIT:

Hier gibt es eine kurze Beschreibung wie man eine UF2 Datei auf den PICO bekommt:

https://www.okdo.com/getting-started/get-started-with-raspberry-pi-pico/

Statt blink.uf2 nimmst du einfach meine Firmware ;-)

**Comment by Helix-FX** [comment]
Ja, das habe ich soweit verstanden mit dem Kopieren. Aber welche Datei kopiere ich auf den Pi? Wie bekomme ich die Daten für Eslcam drauf? Muss die main.c oder ähnliches auch rüber kopiert werden?

**Comment by hdo** [comment]
Ah ok, jetzt verstehe ich was du meinst.

Die Firmware ist bereits fertig kompiliert. Du bekommst die UF2 Datei hier:

https://github.com/hdo/pico-estlcam-handrad/releases/tag/v0.0.1

![image](https://user-images.githubusercontent.com/2122230/148370965-c0b576fc-a522-4bf8-ae97-b1650d1c9466.png)

Du musst also nichts mit den Source selber machen (.c und .h).

Ich nehme an du weißt, dass das Handrad das offizielle Estlcam Protkoll spricht?
D.h. du brauchst entweder den offiziellen LPT/Parallelport Adapter bzgl. den Klemmenadapter.

**Comment by Helix-FX** [comment]
Oh man. Die Datei habe ich vollkommen übersehen. Vielen Dank!

Ja genau, das mit dem I²C Protokoll und Estlcam habe ich alles soweit verstanden. Ich benutze den Klemmenadapter und dementsprechend dann auch den Mini DIN Anschluss.

**Comment by hdo** [comment]
> Oh man. Die Datei habe ich vollkommen übersehen. Vielen Dank!

Manchmal sieht man den Wald vor lauter Bäumen nicht ;-)
 
> Ja genau, das mit dem I²C Protokoll und Estlcam habe ich alles soweit verstanden. Ich benutze den Klemmenadapter und dementsprechend dann auch den Mini DIN Anschluss.

Perfekt. Ich habe den LPT Adapter zur Zeit im Einsatz. Den Klemmen-Adapter habe ich auch da, muss den aber erst noch anschließen und testen :-)

**Comment by Helix-FX** [comment]
Anschließen und Testen steht bei mir die Tage auch an. Ich melde mich mal, wenn ich Erfolg hatte :)

Wenn das wirklich klappt, wäre das der Hammer. Ich habe hier so ein altes Pult von einer CNC Fräse was ich gerne umbauen möchte.
![image](https://user-images.githubusercontent.com/96875040/148380102-a5b737e0-8f93-4683-a511-289dcfb77565.png)

**Comment by hdo** [comment]
Das ist ja ein wuchtiges Teil :-)

**Comment by Helix-FX** [comment]
jo, das will ich dann als Bedienpult direkt an der Fräse montieren :)
Leider habe ich dann noch viele Knöpfe und Potis die nicht belegt sind, aber ich freue mich schon, wenn erstmal der Joystick und ein paar Knöpfe funktionieren. Ich werde wahrscheinlich auch noch irgendwo mittig die Potis wegnehmen und dort dann einen Encoder  drauf setzen. Bestellt ist er schon, leider noch nicht da.

**Comment by hdo** [comment]
Christian hat für Estlcam vor einiger Zeit Makros angekündigt. Ich bin mal gespannt, wann das kommt. Dann kannst du die übrigen Knöpfe dafür nutzen.

Ich arbeite ebenfalls bereits an einer eigenen Makro Lösung :-) Das Auslösen des Werkzeuglängensensors wäre z.B. sehr praktisch.

![2022-01-07_09-10-52](https://user-images.githubusercontent.com/2122230/148512882-98b1c5cd-57f6-4a15-98d1-fd6fbf1afc8a.png)

**Comment by Helix-FX** [comment]
Das wäre richtig gut, wenn das funktioniert!
Sieht gut aus dein Bedienpanel. Ist der Bildschirm mit dem PC verbunden auf dem ESTLCAM läuft?

**Comment by hdo** [comment]
Genau, ursprünglich hatte ich eine Windows 10 Tablet (Chuwi) aber das war nur so mäßig. Dann ein 13.3" Laptop Display und jetzt habe ich ein 13.3 Touch Screen. Das ganze läuft auf jetzt einen Lenovo Tiny PC (M93). Die Kombo ist ziemlich ok, der M93 bootet in 15 Sekunden oder so :-)

**Comment by Helix-FX** [comment]
Das ist cool, ich denke sowas in der Art baue ich mir auch noch. Ein eigenständiges Gerät mit dem ich die Fräse ansteuern kann, auf dem ESTLCAM läuft, was ich aber nicht dauerhaft in der Werkstatt lassen muss wegen Kälte und Dreck. Aber ich bin kein Touchscreenfan....mal schauen was ich mache. Habe noch so einen Home-Theater PC rumliegen. 

Aber erstmal muss die Fräse laufen und dann schaue ich nach und nach weiter.

**Comment by Helix-FX** [comment]
Moin,
ich habe jetzt alles soweit angeschlossen. Zwei Fragen habe ich noch:
1. Ist der Handrad Encoder einfach über GND, 5V VCC, Feed Enc B und Feed Enc A angeschlossen?
2. Wie hast du die Potis ganz unten auf dem Handrad angeschlossen? Sind die auch am Pico Pi?

Schöne Grüße,
Sören

**Comment by hdo** [comment]
Moin Sören,

zu 1) Ja genau, das müssten 4 Pin sein wie du schon korrekt angegeben hast. Ob A oder B, wenn du diese vertauscht ändert sich nur die Richtung

zu 2) Das ist etwas komplizierter. Ich hätte Potis genommen, aber der PICO hat nur 4 Analoge Eingänge, deswegen sind das keine Potis sondern Drehimpulsgeber (Drehencoder). Leider verhalten sich nicht alle Drehencoder gleich. Meine habe ich bei A* bezogen, ich meine in der Stückliste gibt es einen Link dazu.

VG Huy

**Comment by Helix-FX** [comment]
Hi Huy,

ich habe jetzt alles soweit verbunden und die Fräse läuft auch. Nun bekomme ich nur keine Strom auf den Pico Pi. Woher weiß der Klemmenadapter ob der Mini DIN Anschluss in Betriebsart 1 oder 2 läuft? 

LG
Sören

**Comment by hdo** [comment]
Hi Sören,

die 5V kommen vom DIN Anschluss.

![image](https://user-images.githubusercontent.com/2122230/150073320-30a5674d-953f-4613-a67a-d27d6bb046d4.png)

siehe [https://www.estlcam.de/DIN_Detail.php](https://www.estlcam.de/DIN_Detail.php)

Beim PICO schließt du dies an den PIN VCC 5V.

Das ESTLCAM Board sendet eine Anfrage an den PICO und daraus ergibt sich Mode 1 oder 2.

**Comment by Helix-FX** [comment]
Ok, genau das habe ich gemacht. Dann muss bei meiner Verkabelung noch etwas falsch sein, es kommen keine 5V über den Pin 3 an. Ich habe noch eine Buchse und einen Stecker verbaut. Ich vermute dort liegt das Problem. Hatte kurz gedacht, dass man evtl noch in estlcam Konfiguration eingeben muss, dass die Betriebsart 2 genutzt wird. 

Vielen Dank. Und ich melde mich wieder wenn ich es kontrolliert habe 😉

**Comment by sunnysonntag** [comment]
Hallo,

erst einmal danke für die geleistete Arbeit, habe das Handrad nach dieser Anleitung gebaut und bin fast fertig.
Ich kapiere nur nicht wie der Drehencoder, also das große Metallrad mit dem Griff angeschlossen werden muss, wenn ich das so anklemme wie oben beschrieben über GND, 5V VCC, Feed Enc B und Feed Enc A, dann ändert sich im Estlcam der Wert für den Vorschub. Also habe ich an den Pin 24 und 25 den Drehgeber unten links angeklemmt und siehe da, jetzt kann ich den Vorschub damit regeln. Laut Pinout vom PI Pico ist da am Pin 21 und 22 der Spin Encoder B und der Spin Encoder A angeklemmt, ich lese aber in der Beschreibung der Mini Din Buchse des Klemmenadapters, das Drehgeber Encodersignale über zwei Leitungen direkt zur Steuerung übertragen werden. Was kommt jetzt an Pin 21 und 22 ran, auf dem Bild des geöffneten Handrades sieht man da Kabel dran?
So, ich hoffe da blickt noch jemand durch, es wäre super, wenn mir dazu jemand einen Tipp oder Schaltplan oder auch Bilder geben könnte, damit ich das noch hinbekomme!

Gruß
Sunny

**Comment by majooh** [comment]
Hallo,
in der Stückliste fehlt noch das Gegenstück für die Mini DIN Buchse. Was wird den da verwendet?

Gruß Mario
